runtimes/ruby-3.1/server.rb

```diff
@@ -3,117 +3,271 @@
+    else
+      @logs.push(message.to_s)
+    end
   end
 
-  def headers
-    @headers
+  def error(message)
+    if(message.kind_of?(Array) || message.kind_of?(Hash))
+      @errors.push(message.to_json)
+    else
+      @errors.push(message.to_s)
+    end
   end
 end
 
-class RuntimeResponse
-  def send(message)
-    @response = message
+def handle(request, response)
+  secret = request.env['HTTP_X_OPEN_RUNTIMES_SECRET'] || ''
+  serverSecret = ENV['OPEN_RUNTIMES_SECRET'] || ''
+
+  if(secret == '' || secret != serverSecret)
+    response.status = 500
+    response.body = 'Unauthorized. Provide correct "x-open-runtimes-secret" header.'
+    return
+  end
+
+  request.body.rewind
+
+  host = request.host
+
+  scheme = request.scheme || 'http'
+  defaultPort = scheme === 'https' ? '443' : '80'
+  port = request.port || defaultPort.to_i
+  path = request.path
+  query = {}
+  queryString = request.query_string || ''
+
+  if(queryString.start_with?("?"))
+    queryString = queryString[1..-1]
   end
 
-  def json(object)
-    @response = object
+  url = scheme + "://" + host
+
+  if(port != defaultPort.to_i)
+    url += ':' + port.to_s
   end
-end
 
-post '/' do
-  challenge = request.env['HTTP_X_INTERNAL_CHALLENGE'] || ''
+  url += path
 
-  if challenge == ''
-    status 500
-    return { stderr: 'Unauthorized' }.to_json
+  if(!(queryString.empty?))
+    url += "?" + queryString
   end
 
-  if challenge != ENV['INTERNAL_RUNTIME_KEY']
-    status 500
-    return { stderr: 'Unauthorized' }.to_json
+
+  if(!(queryString.empty?))
+    queryString.split('&') do |param|
+      pair = param.split('=', 2)
+      if(pair[0] != nil && !(pair[0].empty?))
+        query[pair[0]] = pair[1]
+      end
+    end
   end
 
+  bodyString = request.body.read
+  body = bodyString
+  method = request.request_method
+  headers = {}
 
-  request.body.rewind
-  data = JSON.parse(request.body.read)
+  if(request.env['CONTENT_TYPE'] != nil)
+    headers['content-type'] = request.env['CONTENT_TYPE']
+  end
 
-  requestData = RuntimeRequest.new(data['payload'], data['variables'], data['headers'])
-  runtimeResponse = RuntimeResponse.new
+  if(request.env['CONTENT_LENGTH'] != nil)
+    headers['content-length'] = request.env['CONTENT_LENGTH']
+  end
 
-  begin
-    load(USER_CODE_PATH + '/' + ENV['INTERNAL_RUNTIME_ENTRYPOINT'])
-  rescue Exception => e
-    p e
-    status 500
-    return {stderr: e.backtrace.join("\n") }.to_json
+  request.env.each do |header, value|
+    if(header.start_with?('HTTP_'))
+      header = header[5..-1]
+      header = header.gsub("_", "-")
+      header = header.downcase
+
+      if(!header.start_with?('x-open-runtimes-'))
+        headers[header] = value
+      end
+    end
   end
 
-  unless defined?(main = ())
-    status 500
-    return { stderr: 'File does not specify a main() function.' }.to_json
+  contentType = request.env['CONTENT_TYPE']
+  if(contentType == nil)
+    contentType = 'text/plain'
   end
```

Comment:
You could use a trailing if here:

```suggestion
content_type = 'text/plain' if content_type.nil?
```

Reply:
Appleid here and in other places as well.